Bottom: 5443b430419fc07b750ee98358ff646fef9e996a
Top:    af783a8e21b42ceab0ffb5a59dd643d10d4c11da
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-12-16 15:29:34 +0800

XDP: introduce XDP_HOLD


---

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_rx.c b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
index 9fea7e4..87c314c 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_rx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
@@ -916,6 +916,8 @@ int mlx4_en_process_rx_cq(struct net_device *dev, struct mlx4_en_cq *cq, int bud
 			switch (act) {
 			case XDP_PASS:
 				break;
+			case XDP_HOLD:
+				goto next;
 			case XDP_TX:
 				if (likely(!mlx4_en_xmit_frame(ring, frags, dev,
 							length, cq->ring,
diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
index 7d9b283..4e177aa 100644
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -530,6 +530,7 @@ enum xdp_action {
 	XDP_DROP,
 	XDP_PASS,
 	XDP_TX,
+	XDP_HOLD,
 };
 
 /* user accessible metadata for XDP packet hook
